src/backend/ingestion/framework-detection.lambda-shared.ts

```diff
@@ -28,65 +40,67 @@ export interface ConstructFramework {
+  // e.g. in the case of a transitive dependency where the info isn't provided
+  // "ambiguous" means we have seen multiple major versions so it's impossible
+  // to resolve a single number
+  const detectedFrameworks: { [P in ConstructFrameworkName]?: number | 'not-sure' | 'ambiguous' } = {};
 
   detectConstructFrameworkPackage(assembly.name, assembly.version);
   for (const depName of Object.keys(assembly.dependencyClosure ?? {})) {
     detectConstructFrameworkPackage(depName);
-    if (nameAmbiguous) {
-      return undefined;
+  }
+
+  const frameworks: ConstructFramework[] = [];
+  for (const [frameworkName, majorVersion] of Object.entries(detectedFrameworks)) {
+    const name = frameworkName as ConstructFrameworkName;
+    if (majorVersion === undefined) {
+      continue;
+    } else if (majorVersion === 'ambiguous' || majorVersion === 'not-sure') {
+      frameworks.push({ name });
+    } else if (typeof majorVersion === 'number') {
+      frameworks.push({ name, majorVersion });
     }
   }
 
-  return name && { name, majorVersion: majorVersionAmbiguous ? undefined : majorVersion };
+  return frameworks;
 
+  /**
+   * Analyses the package name and version range, and updates
+   * `detectedFrameworks` from the parent scope appropriately
+   */
   function detectConstructFrameworkPackage(packageName: string, versionRange = assembly.dependencies?.[packageName]): void {
-    if (packageName.startsWith('@aws-cdk/') || packageName === 'aws-cdk-lib' || packageName === 'monocdk') {
-      if (name && name !== ConstructFrameworkName.AWS_CDK) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.AWS_CDK;
-    } else if (packageName === 'cdktf') {
-      if (name && name !== ConstructFrameworkName.CDKTF) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.CDKTF;
-    } else if (packageName === 'cdk8s' || /^cdk8s-plus(?:-(?:17|20|21|22))?$/.test(packageName)) {
-      if (name && name !== ConstructFrameworkName.CDK8S) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.CDK8S;
-    } else {
-      return;
-    }
-    if (versionRange) {
-      const major = minVersion(versionRange)?.major;
-      if (majorVersion != null && majorVersion !== major) {
-        // Identified multiple candidates, so this is ambiguous...
-        majorVersionAmbiguous = true;
+    for (const frameworkName of [ConstructFrameworkName.AWS_CDK, ConstructFrameworkName.CDK8S, ConstructFrameworkName.CDKTF]) {
+      const matchesFramework = FRAMEWORK_MATCHERS[frameworkName];
+      if (matchesFramework(packageName)) {
+        const frameworkVersion: number | 'not-sure' | 'ambiguous' | undefined = detectedFrameworks[frameworkName];
```

Comment:
Is the type annotation _really_ necessary here?

```suggestion
        const frameworkVersion = detectedFrameworks[frameworkName];
```